### PR TITLE
Text classification for Scikit-learn

### PR DESCRIPTION
**Pull request by merveenoyan**
@BenjaminBossan @adrinjalali I've written main pipeline. For tests, I don't know what type of parameterization we need to do compared to tabular cases (probably not much?)

Also here, I checked if labels are given in the config (we currently don't have it) and if not, we take them from `model.classes_`. Does this sound good?

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
> For tests, I don't know what type of parameterization we need to do compared to tabular cases (probably not much?)

Did you see the documentation for testing in the [README](https://github.com/huggingface/api-inference-community/blob/main/docker_images/sklearn/README.md)? I think this should give a good indication about how to test.

As for the models themselves (not sure if you refer to that), I would probably stick close to the existing ones. So logistic regression and hist gradient boosting, of course with something like tfidf as a vectorization step. It looks like HGB does not support sparse matrices, so that would require casting to dense as an intermediate step (I guess through `FunctionTransformer`).

Regarding the implementation, I think with a tiny bit of refactoring, we can avoid a lot of code duplication in the Pipeline implementation:

If inside of `TabularClassificationPipeline.__init__`, the config would be stored on self, you could do:

```python
class TextClassificationPipeline(TabularClassificationPipeline):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # use labels from config file if available
        self.labels = self.config.get("sklearn", {}).get("labels", None)
        if not self.labels:
            self.labels = self.model.classes_
```

(btw. @adrinjalali can we assume that `model.classes_` always exists?)

Similarly, with a little bit of refactoring, we can avoid the duplication within `__call__`. For this, I would move the only lines of code that differ -- those relating to creating the output from the prediction -- into a submethod, say `_get_model_output`. Then you only need to override that and don't need to copy all the error handling stuff.

**Comment by adrinjalali** [comment]
I think we can start with the same usecases we have for tabular-classification, but with text input.

**Comment by merveenoyan** [comment]
@adrinjalali @BenjaminBossan I'm checking for old/new sklearn version, `has_config` and only one input output.

**Comment by merveenoyan** [comment]
@BenjaminBossan I wonder why we should have both `LogisticRegression` and `HistGradientBoostingClassifier` (referring to what you said in slack) given for tabular tasks you can have missing values so it's good to have one that can't handle missing values and another that can handle missing values, but I don't see a need for test case on text classification. Instead, we could have only `HistGradientBoostingClassifier` across different sklearn versions, no?

File: docker_images/sklearn/app/pipelines/text_classification.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import json
```
Comment (adrinjalali) [inline docker_images/sklearn/app/pipelines/text_classification.py:1]: I think we should refactor so that this file doesn't repeat as much code as it is right now from the `tabular-classification` task.

File: docker_images/sklearn/tests/test_api.py [hunk 2]
```diff
@@ -120,15 +120,15 @@
             "has_config": False,
             "old_sklearn": True,
             "accepts_nan": False,
-            "loads_on_new_sklearn": True,
+            "loads_on_new_sklearn": False,
```
Comment (adrinjalali) [inline docker_images/sklearn/tests/test_api.py:123]: why is this being modified?

**Review by adrinjalali — COMMENTED** [review]
Would be nice to also run black on the new scripts.

**Comment by merveenoyan** [comment]
Please do not review, it's a draft PR, there's a lot I have to do.

**Comment by BenjaminBossan** [comment]
> @BenjaminBossan I wonder why we should have both `LogisticRegression` and `HistGradientBoostingClassifier`

The question was answered on slack, right?

**Comment by merveenoyan** [comment]
@BenjaminBossan not that one, given I wanted it to be transparent I asked from here :')

**Comment by merveenoyan** [comment]
I tested it on my local, it works, I'll do refactors you've asked though but you can review @adrinjalali @BenjaminBossan

**Comment by merveenoyan** [comment]
I added a custom transformer as `HistGradientBoostingRegressor` doesn't accept `CountVectorizer` output (sparse matrix). It also takes a lot of time on CPU, I wonder if it's actually expected or if it's acceptable to make the model faster.

**Comment by BenjaminBossan** [comment]
> I added a custom transformer as `HistGradientBoostingClassifier` doesn't accept `CountVectorizer` output (sparse matrix).

I guess a `FunctionTransformer` would have worked as well in case you want to avoid a custom transformer.

> It also takes a lot of time on CPU

I think there are two reasons: First the dataset is quite a bit bigger than the other ones. Second, you don't limit the vocab size, leading to a huge (dense) matrix. I'd recommend setting `max_features` to something small like 100 or so -- the accuracy is not important here.

> I was wondering if assertion to an expected output makes sense as we do `predict_proba`, I'm wondering if outputs might be probabilistic. (not sure)

If I understand you correctly, yes, it should work (maybe requiring `random_state` being fixed), at least up to a certain precision.

**Comment by merveenoyan** [comment]
@BenjaminBossan yes I meant that there's a randomness introduced. Maybe it's better if I round the probabilities when comparing. 
As for dataset I thought it's a toy dataset so I didn't think it would be worse 😂 makes a lot of sense! Logistic regression takes sparse matrix so it's easy.

**Comment by Narsil** [comment]
@merveenoyan any real model on hf.co we can try on ? (I would like to add it to `tests/test_dockers.py` which is the integration test here.

**Comment by merveenoyan** [comment]
@Narsil https://huggingface.co/merve/20newsgroups
The only thing that's left is fixing the tests (we couldn't decide if we should pre/post-process labels til yesterday) so I'm doing that.

**Comment by Narsil** [comment]
Seems to be working. I'm off next week, but LGTM

**Comment by merveenoyan** [comment]
@BenjaminBossan @adrinjalali I edited the tests and they're passing now. My only problem is that the sklearn version warning test raises the error as intended and assertion is done but it acts as if it's not working. I don't get that. 

Can you make a sanity check on how the `test_success_code` takes data, the data generated and how the main app takes data? Does below seem good? 
See request below:
```python
(api_inf) ➜  generators git:(text_clf_sklearn) ✗ curl -X POST -d 'this movie is cool' http://localhost:8000/
[[{"label":"0","score":0.18274385806392676},{"label":"1","score":0.3020991751096761},{"label":"2","score":0.23591785291978498},{"label":"3","score":0.27923911390661116}]]%    
```
The successful test:
```python
def test_success_code(self):
        self._check_requirement(not self.case_data["old_sklearn"])
        
        data = self.data
        expected_output_len = len(self.expected_output)
        
        with TestClient(self.app) as client:
            
            response = client.post("/", json={"inputs": data["data"][0]})

        self.assertEqual(
            response.status_code,
            200,
        )
```
I was wondering, it takes string as text, but what if the user enters list of texts? I know it's the way pipeline works but I don't know if it sounds right. 

```python
def __call__(self, inputs: str) -> List[Dict[str, float]]:
        """
        Args:
            inputs (:obj:`str`):
                a string containing some text
        Return:
            A :obj:`list`:. The object returned should be a list of one list like [[{"label": 0.9939950108528137}]] containing:
                - "label": A string representing what the label/class is. There can be multiple labels.
                - "score": A score between 0 and 1 describing how confident the model is for this label/class.
        """
        if self._load_exception:
            raise ValueError(
                "An error occurred while loading the model: "
                f"{str(self._load_exception)}"
            )

        _warnings = []
        exception = None
        try:
            with warnings.catch_warnings(record=True) as record:
                res = []
                for i, c in enumerate(self.model.predict_proba([inputs]).tolist()[0]):
                    res.append({"label": str(self.model.classes_[i]), "score": c})
                res = [res]
```

**Comment by merveenoyan** [comment]
@adrinjalali @BenjaminBossan 
pw5cp1kz4hi41

File: docker_images/sklearn/tests/test_api.py (hunk 2 above)
Comment (adrinjalali) [reply docker_images/sklearn/tests/test_api.py:133]: this is not answered

**Review by adrinjalali — COMMENTED** [review]
Regarding inputs, I do agree that accepting a list of strings would make sense, rather than having to send a separate request for each input. But since this is a much more established part of the API and all other libraries use it, we need to be careful touching it. For now we can do as other libraries do, but we should probably think about it. WDYT @osanseviero ?

As for the tests, I'm wondering if we should also refactor tests, at this point we have 3 copies of the tests, and they might slightly differ.

I think the code for the pipeline itself is also duplicated, we should refactor the pipelines and extract common code from them, WDYT @BenjaminBossan ?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for working on this Merve, most of the content already looks very good to me. I have, however, a few comments, please take a look.

On top of those comments:

> I was wondering, it takes string as text, but what if the user enters list of texts? I know it's the way pipeline works but I don't know if it sounds right.

I'm not 100% sure I understand you correctly, but it looks like the current implementation only works if the user passes a single sample, whereas for the existing pipelines, it's possible to pass multiple samples. Therefore, the code should be changed here to support multiple samples.

It also seems to me that the unit tests have to be expanded a bit. If I'm not missing something, you never test the actual model predictions, so they could be completely wrong. The existing tests do check the model output, so maybe you can take some of those checks and add them here.

**Comment by osanseviero** [comment]
> Regarding inputs, I do agree that accepting a list of strings would make sense, rather than having to send a separate request for each input. But since this is a much more established part of the API and all other libraries use it, we need to be careful touching it. For now we can do as other libraries do, but we should probably think about it. WDYT @osanseviero ?

We have supported batching inputs for `transformers` for quite a while, but we did a decision last year to not support batching in community at the moment. Iirc it should be relatively easy to add support for it (https://github.com/huggingface/api-inference-community/issues/31), but I think there were some arguments against it. @Narsil can probably share more on it.

File: docker_images/sklearn/app/pipelines/text_classification.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply docker_images/sklearn/app/pipelines/text_classification.py:1]: I don't see it anywhere?

[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply docker_images/sklearn/app/pipelines/text_classification.py:1]: Sorry, my bad. I was looking at the file names in `common/app/pipeline`:

- https://github.com/huggingface/api-inference-community/blob/main/docker_images/common/app/pipelines/tabular_classification_pipeline.py
- https://github.com/huggingface/api-inference-community/blob/main/docker_images/common/app/pipelines/tabular_regression_pipeline.py

**Comment by merveenoyan** [comment]
Creating a child class that is inheriting a parent class that inherits `TestCase` is incredibly tricky. The biggest tricky part is that functions in parent class doesn't receive parameters given with `parameterized_class`. (This is a note to future me, we looked into it with Benjamin and the tests are trying everything to resist to working lol)
I also tried using dummy values so it could at least assume variables like `test_case` are class attributes and receive from `parameterized_class`, this time the child specific tests pass but parent specific tests fail. (they also fail when I don't give a dummy)
So I gave up on refactoring and will cherry pick a commit I'll go back to and just merge text-classification here.

**Comment by merveenoyan** [comment]
@adrinjalali @BenjaminBossan I reverted my refactoring commit and made the changes you asked.

File: docker_images/sklearn/app/pipelines/text_classification.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    open(Path(cached_folder) / self.model_file, "rb")
+                )
+                if len(record) > 0:
+                    # if there's a warning while loading the model, we save it so
+                    # that it can be raised to the user when __call__ is called.
+                    self._load_warnings += record
+        except Exception as e:
+            # if there is an exception while loading the model, we save it to
+            # raise the write error when __call__ is called.
+            self._load_exception = e
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docker_images/sklearn/app/pipelines/text_classification.py:52]: this is not addressed @merveenoyan

**Comment by merveenoyan** [comment]
@adrinjalali @BenjaminBossan I regenerated everything and tested for I refactored everything by creating a new sklearn pipeline. Every test passes.

File: docker_images/sklearn/tests/generators/samples/iris-1.0.0-input.json [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docker_images/sklearn/tests/generators/samples/iris-1.0.0-input.json:1]: I think the 1.0.0 files were added since @merveenoyan called the generator directly instead of using `run.sh`, these files should be removed.

File: docker_images/sklearn/tests/test_api.py [hunk 6]
```diff
@@ -128,7 +138,7 @@
             "has_config": True,
             "old_sklearn": True,
             "accepts_nan": False,
-            "loads_on_new_sklearn": True,
+            "loads_on_new_sklearn": False,
```
Comment (adrinjalali) [inline docker_images/sklearn/tests/test_api.py:141]: why is this changed?

**Comment by merveenoyan** [comment]
@adrinjalali sorry I didn't do nitpicking & bignitpicking, I was a bit tired so I'll check everything out and ask for a review.

**Review by adrinjalali — APPROVED** [review]
Other than @BenjaminBossan 's comments LGTM.

**Review by BenjaminBossan — APPROVED** [review]
Overall very good, happy to see this come to a finish.

As I have mentioned previously, I think that the `__call__` method of the different pipelines could be simplified, so that we don't have that much duplicated code in `TextClassificationPipeline.__call__`. However, I would be okay with making that change at a later point in time.

**Comment by merveenoyan** [comment]
@Narsil can you review? 🥺

**Review by Narsil — APPROVED** [review]
LGTM.

Maybe let's just add a real model to the `tests/test_docker` to make sure everything will be working fine.

**Comment by merveenoyan** [comment]
@Narsil what do you mean? We have a lot of models with expected input outputs inside the repository that we check with.

**Comment by merveenoyan** [comment]
@Narsil I debugged sklearn tests on my local, this model: https://huggingface.co/skops-tests/iris-sklearn-latest-logistic_regression-with-config didn't have any task associated so I fixed it, now running tests on my local. will ask for a re-review if they pass.

**Comment by merveenoyan** [comment]
I infer with the model just fine: 

```
(api_inf) ➜  api-inference-community git:(text_clf_sklearn) python3 ./manage.py start skops-tests/iris-sklearn-latest-logistic_regression-with-config
                
Inferred task : tabular-classification
Inferred framework : sklearn
INFO:     Started server process [6317]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
2022-09-20 15:23:00,731 - INFO - 127.0.0.1:51577 - "POST / HTTP/1.1" 200
```
The request:
```
(api_inf) ➜  api-inference-community git:(text_clf_sklearn) curl -X POST -d'{"data":{"sepal length (cm)":[6.1, 5.7, 7.7],"sepal width (cm)":[2.8, 3.8, 2.6],"petal length (cm)":[4.7, 1.7, 6.9],"petal width (cm)":[1.2, 0.3, 2.3]}}' http://localhost:8000/

[1,0,2]% 
```
I don't know why in tests I get timeout for some reason :( @Narsil

**Comment by Narsil** [comment]
Yes, we removed those automated tests because they were too slow.

First, there's  `mamba` being used to create the environment at startup. This too slow for one part of the test that checks that secondary loads are faster (usually because the model is already downloaded and on the machine).

Here the mamba creation is kind of incompressible so too slow and this tests fails. It's OK since it's the best solution we have to run sklearn.

I'm trying to figure out but it seems there's another part of the tests which fails which is that the docker is not stopping properly.
This is slightly more worrysome (nothing huge we can merge without it, as kubernetes as other mecanisms in place to cope, but it would be nice to fix)
